### PR TITLE
ME i/o hatches for multiblocks

### PR DESCRIPTION
**Pull request by screret**
Pattern-providing ones _might_ come in a separate PR.

### CODE REVIEW

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/gui/widget/AEConfigWidget.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    public final IConfigurableSlot getConfig(int index) {
+        return this.config[index];
+    }
+
+    public final IConfigurableSlot getDisplay(int index) {
+        return this.displayList[index];
+    }
+
+    protected final boolean areAEStackCountEquals(GenericStack s1, GenericStack s2) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/gui/widget/AEConfigWidget.java]: Small typo: should be `areAEStackCountsEqual`

**Comment by screret** [comment]
oh fuck, merge conflicts spawned

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEInputBusPartMachine.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    protected NotifiableItemStackHandler createInventory(Object... args) {
+        this.aeItemHandler = new ExportOnlyAEItemList(this, CONFIG_SIZE);
+        return this.aeItemHandler;
+    }
+
+    @Override
+    public void autoIO() {
+        if (getLevel().isClientSide) return;
+        if (this.workingEnabled && this.shouldSyncME()) {
+            if (this.updateMEStatus()) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEInputBusPartMachine.java]: I'd use two early returns here, to reduce nesting depth

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEInputHatchPartMachine.java [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+        modularUI.widget(UITemplate.bindPlayerInventory(entityPlayer.getInventory(), GuiTextures.SLOT, 7, 18 + 18 * 4 + 12, true));
+        return modularUI;
+    }
+
+    @Override
+    protected void autoIO() {
+        if (getLevel().isClientSide) return;
+        if (this.workingEnabled && this.shouldSyncME()) {
+            if (this.updateMEStatus()) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEInputHatchPartMachine.java]: Early returns would be good here as well

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEOutputBusPartMachine.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    protected NotifiableItemStackHandler createInventory(Object... args) {
+        this.internalBuffer = new SerializableGenericStackInv(this::onChanged, 16);
+        return new InaccessibleInfiniteSlot(this, this.internalBuffer);
+    }
+
+    @Override
+    public void autoIO() {
+        if (getLevel().isClientSide) return;
+        if (this.workingEnabled && this.shouldSyncME()) {
+            if (this.updateMEStatus()) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEOutputBusPartMachine.java]: Here as well

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEOutputHatchPartMachine.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+        modularUI.widget(UITemplate.bindPlayerInventory(entityPlayer.getInventory(), GuiTextures.SLOT, 7, 18 + 18 * 4 + 12, true));
+        return modularUI;
+    }
+
+    @Override
+    protected void autoIO() {
+        if (getLevel().isClientSide) return;
+        if (this.workingEnabled && this.shouldSyncME()) {
+            if (this.updateMEStatus()) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEOutputHatchPartMachine.java]: And here as well

**Review by mikerooni — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/util/AEFluidConfigSlot.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+public class AEFluidConfigSlot extends AEConfigSlot {
+    public static final int LOAD_PHANTOM_FLUID_STACK_FROM_NBT = 13;
+
+
+    public AEFluidConfigSlot(int x, int y, AEConfigWidget widget, int index) {
+        super(new Position(x, y), new Size(18, 18 * 2), widget, index);
+    }
+
+    @Override
+    public void drawInBackground(@NotNull GuiGraphics graphics, int mouseX, int mouseY, float partialTicks) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/util/AEFluidConfigSlot.java:51]: Might be able to extract some parts of this into `AEConfigSlot` (optional)

File: common/src/main/java/com/gregtechceu/gtceu/integration/ae2/util/AEFluidConfigSlot.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    private int tryClickContainer(boolean isShiftKeyDown) {
+        MEInputHatchPartMachine.ExportOnlyAEFluid fluidTank = this.parentWidget.getConfig(this.index) instanceof MEInputHatchPartMachine.ExportOnlyAEFluid fluid ? fluid : null;
+        if (fluidTank == null) return -1;
+        Player player = gui.entityPlayer;
+        ItemStack currentStack = gui.getModularUIContainer().getCarried();
+        var handler = FluidTransferHelper.getFluidTransfer(gui.entityPlayer, gui.getModularUIContainer());
+        if (handler == null) return -1;
+        int maxAttempts = isShiftKeyDown ? currentStack.getCount() : 1;
+
+        if (fluidTank.getFluidAmount() > 0) {
```
Comment (mikerooni) [inline common/src/main/java/com/gregtechceu/gtceu/integration/ae2/util/AEFluidConfigSlot.java:264]: might be better to early return here

**Review by mikerooni — APPROVED** [review]

**Pull request merged**